File: 291_word_pattern_ii.rb
=begin
Given a pattern and a string str, find if str follows the same pattern.

Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty substring in str.

Example 1:

Input: pattern = "abab", str = "redblueredblue"
Output: true
Example 2:

Input: pattern = pattern = "aaaa", str = "asdasdasdasd"
Output: true
Example 3:

Input: pattern = "aabb", str = "xyzabcxzyabc"
Output: false
Notes:
You may assume both pattern and str contains only lowercase letters.
=end

require 'set'
# @param {String} pattern
# @param {String} str
# @return {Boolean}
def word_pattern_match(pattern, str)
    map = {}
    set = Set.new
    is_match(pattern, str, 0, 0, map, set)
end

def is_match(pattern, str, p, s, map, set)
    if p == pattern.length && s == str.length
        return true
    end
    if p == pattern.length || s == str.length
        return false
    end
    
    char = pattern[p]
    if map[char]
        word = map[char]
        if word != str[s...s+word.length]
            return false
        end
        
        return is_match(pattern, str, p+1, s+word.length, map, set)
    end
    
    (s...str.length).each do |i|
        word = str[s..i]
        if set.include?(word)
            next
        end
        
        map[char] = word
        set.add(word)
        if is_match(pattern, str, p+1, i+1, map, set)
            return true
        end
        map.delete(char)
        set.delete(word)
    end
    
    return false
end

File: 291_word_pattern_ii.py
"""
Given a pattern and a string str, find if str follows the same pattern.

Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty substring in str.

Example 1:

Input: pattern = "abab", str = "redblueredblue"
Output: true
Example 2:

Input: pattern = pattern = "aaaa", str = "asdasdasdasd"
Output: true
Example 3:

Input: pattern = "aabb", str = "xyzabcxzyabc"
Output: false
Notes:
You may assume both pattern and str contains only lowercase letters.
"""

class Solution:
    def wordPatternMatch(self, pattern: str, str: str) -> bool:
        wordByChar = {}
        seen = set()
        return self.isMatch(0, 0, pattern, str, wordByChar, seen)
    
    def isMatch(self, p, s, pattern, str, wordByChar, seen):
        if len(pattern) == p and len(str) == s:
            return True
        if len(pattern) == p or len(str) == s:
            return False
        char = pattern[p]
        if char in wordByChar:
            word = wordByChar[char]
            if word != str[s:s+len(word)]:
                return False
            
            return self.isMatch(p+1, s+len(word), pattern, str, wordByChar, seen)
        
        for i in range(s, len(str)):
            word = str[s:i+1]
            if word in seen:
                continue
            wordByChar[char] = word
            seen.add(word)
            if self.isMatch(p+1, i+1, pattern, str, wordByChar, seen):
                return True
            del wordByChar[char]
            seen.remove(word)
        
        return False
        
